deepchem/models/torch_models/kfac_optimizer.py

```diff
@@ -0,0 +1,430 @@
+
+  def ComputeCovG(self, g: torch.Tensor,
+                  layer: torch.nn.Module) -> torch.Tensor:
+    """
+    Compute the covariance matrix of the G matrix (the gradient of the layer).
+    Parameters:
+    -----------
+    g: torch.Tensor
+    It is the gradient of the layer for which the covariance matrix should be calculated.
+    layer: torch.nn.Module
+    It specifies the type of layer from which the output of the layer is taken.
+    Returns:
+    --------
+    torch.Tensor
+    The covariance matrix of the G matrix.
+    """
+    if isinstance(layer, torch.linear):
+      batch_size = g.size(0)
+      if self.batch_averaged:
+        cov_g = g.t() @ (g * batch_size)
+      else:
+        cov_g = g.t() @ (g / batch_size)
+      return cov_g
+
+    elif isinstance(layer, torch.Conv2d):
+      spatial_size = g.size(2) * g.size(3)
+      batch_size = g.shape[0]
+      g = g.transpose(1, 2).transpose(2, 3)
+      g = self.try_contiguous(g)
+      g = g.view(-1, g.size(-1))
+      if self.batch_averaged:
+        g = g * batch_size
+      g = g * spatial_size
+      cov_g = g.t() @ (g / g.size(0))
+
+      return cov_g
+
+  def _save_input(self, module: torch.nn.Module, input: torch.Tensor):
+    """
+    Updates the input of the layer using exponentially weighted averages of the layer input.
+    Parameters:
+    -----------
+    module: torch.nn.Module
+    specifies the layer for which the input should be taken
+    input: torch.Tensor
+    the input matrix which should get updated
+    """
+    if torch.is_grad_enabled() and self.steps % self.TCov == 0:
+      aa = self.ComputeCovA(input[0].data, module)
+      # Initialize buffers
+      if self.steps == 0:
+        self.m_aa[module] = torch.diag(aa.new(aa.size(0)).fill_(1))
+      self.m_aa[module] *= self.stat_decay + aa * (1 - self.stat_decay)
+
+  def _save_grad_output(self, module: torch.nn.Module, grad_input: torch.Tensor,
+                        grad_output: torch.Tensor):
+    """
+    Updates the backward gradient of the layer using exponentially weighted averages of the layer input.
+    Parameters:
+    -----------
+    module: torch.nn.Module
+    specifies the layer for which the gradient should be taken
+    input: torch.Tensor
+    the gradient matrix which should get updated
+    """
+    # Accumulate statistics for Fisher matrices
+    if self.acc_stats and self.steps % self.TCov == 0:
+      gg = self.ComputeCovG(grad_output[0].data, module, self.batch_averaged)
+      # Initialize buffers
+      if self.steps == 0:
+        self.m_gg[module] = torch.diag(gg.new(gg.size(0)).fill_(1))
+      self.m_gg[module] *= self.stat_decay + gg * (1 - self.stat_decay)
+
+  def _prepare_model(self):
+    """"
+    Attaches hooks(saving the ouptut and grad according to the update function) to the model for
+    to calculate gradients at every step.
+    """
+    count = 0
+    print(self.model)
+    print("=> We keep following layers in KFAC. ")
```

Comment:
Logging should be done with the `logging` package rather than `print()` calls.  But is this really intended to be logged, or is it just debugging code that got left in?